core-bundle/contao/drivers/DC_Table.php

```diff
@@ -3205,109 +3205,105 @@ protected function save($varValue)
+						}
+						elseif (\is_callable($callback))
+						{
+							$arrValues = $callback($arrValues, $this);
+						}
 					}
-					elseif (\is_callable($callback))
+					catch (\Exception $e)
 					{
-						$arrValues = $callback($arrValues, $this);
+						$this->noReload = true;
+						Message::addError($e->getMessage());
+
+						break;
 					}
 
 					if (!\is_array($arrValues))
 					{
 						throw new \RuntimeException('The onbeforesubmit_callback must return the values!');
 					}
 				}
-				catch (\Exception $e)
-				{
-					$this->noReload = true;
-					Message::addError($e->getMessage());
-
-					return;
-				}
 			}
 		}
 
-		$arrTypes = array();
-		$blnVersionize = false;
+		// Check permissions
 		$currentRecord = $this->getCurrentRecord();
 
-		foreach ($arrValues as $strField => $varValue)
+		$this->denyAccessUnlessGranted(ContaoCorePermissions::DC_PREFIX . $this->strTable, new UpdateAction($this->strTable, $currentRecord, $arrValues));
+
+		// Persist values
+		if (!$this->noReload && !empty($arrValues))
 		{
-			$this->strField = $strField;
-			$arrData = $GLOBALS['TL_DCA'][$this->strTable]['fields'][$this->strField] ?? array();
+			$arrTypes = array();
+			$blnVersionize = false;
 
-			// If the field is a fallback field, empty all other columns (see #6498)
-			if ($varValue && ($arrData['eval']['fallback'] ?? null))
+			foreach ($arrValues as $strField => $varValue)
 			{
-				$varEmpty = Widget::getEmptyValueByFieldType($GLOBALS['TL_DCA'][$this->strTable]['fields'][$this->strField]['sql'] ?? array());
-				$arrType = array_filter(array($GLOBALS['TL_DCA'][$this->strTable]['fields'][$this->strField]['sql']['type'] ?? null));
+				$arrData = $GLOBALS['TL_DCA'][$this->strTable]['fields'][$strField] ?? array();
 
-				if (($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['mode'] ?? null) == self::MODE_PARENT)
-				{
-					$this->Database
-						->prepare("UPDATE " . $this->strTable . " SET " . Database::quoteIdentifier($this->strField) . "=? WHERE pid=?")
-						->query('', array($varEmpty, $currentRecord['pid'] ?? null), $arrType);
-				}
-				else
+				// If the field is a fallback field, empty all other columns (see #6498)
+				if ($varValue && ($arrData['eval']['fallback'] ?? null))
 				{
-					$this->Database
-						->prepare("UPDATE " . $this->strTable . " SET " . Database::quoteIdentifier($this->strField) . "=?")
-						->query('', array($varEmpty), $arrType);
-				}
-			}
-
-			$arrTypes[] = $GLOBALS['TL_DCA'][$this->strTable]['fields'][$this->strField]['sql']['type'] ?? null;
-
-			if (!isset($arrData['eval']['versionize']) || $arrData['eval']['versionize'] !== false)
-			{
-				$blnVersionize = true;
-			}
+					$varEmpty = Widget::getEmptyValueByFieldType($GLOBALS['TL_DCA'][$this->strTable]['fields'][$strField]['sql'] ?? array());
+					$arrType = array_filter(array($GLOBALS['TL_DCA'][$this->strTable]['fields'][$strField]['sql']['type'] ?? null));
 
-			$this->varValue = StringUtil::deserialize($varValue);
```

Comment:
It seems like this line is missing in the new code. Did you remove it on purpose?